gpu/octree/src/cuda/radius_search.cu

```diff
@@ -260,7 +260,13 @@ namespace pcl
                     if (test)
                     {                                        
                         float3 active_query;
+                        float active_radius;
 
+                        //broadcast warp_radius
+                        if (active_lane == laneId)
```

Comment:
I feel like this needs sync threads before load.

Reply:
I believe the `__any_sync` primitive in line 258 will perform the syncronization here, and since we're getting the radius here using essentially the same method used to get the query below, I don't think additional synchronization is necessary. But I'm still learning about this as well so I definitely can't say for certain.

Reply:
> since we're getting the radius here using essentially the same method used to get the query below, I don't think additional synchronization is necessary

This assumes the original isn't buggy. Code can have bugs and still give correct behavior (usually when UB happens)

> `__any_sync` primitive in line 258 will perform the syncronization here

`__any_sync` isn't a sync primitive, but a test primitive. From CUDA docs:
```
__any_sync(unsigned mask, predicate):
    Evaluate predicate for all non-exited threads in mask and return non-zero if and only if predicate evaluates to non-zero for any of them.
```

If this actually resolves the issue, I'll approve this, but I feel something stinky is going on.

Reply:
> __any_sync isn't a sync primitive, but a test primitive. From CUDA docs:

I think when cuda 9.0 came out, they changed these primitives (like __any or __ballot) to their sync variants (__any_sync , __ballot_sync). From their guide [here](https://devblogs.nvidia.com/using-cuda-warp-level-primitives/)
 `All the participating threads must be synchronized for the collective operation to work correctly. Therefore, these primitives first synchronize the threads if they are not already synchronized.`

Reply:
The threads **are synced** for testing the primitive. But after that any operation (not just a load/store) will unsync them. The distance isn't large from sync point, so it's a possibility the sync-load doesn't impact value (or impacts only a little since the algorithm is iterative). I'm basing my conclusions only on the details in [CUDA programming guide (§5.4.2)](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#control-flow-instructions) and experience with CPU atomics.

Again, I'm making noise based on theory and limited knowledge.

Reply:
> But after that any operation (not just a shared mem load/store) will unsync them

Yeah, I'm certainly not too certain about this either, my uunderstanding was that untill the threads perform certain thread specific tasks that could make them divergent, they would remain in sync (such as in listing 13 in the previous link I shared). But I haven't worked with warp level programming at all before, so I really don't have the knowledge to back that up, and might easily be making a wrong assumption.

Reply:
I think listing 9 is more applicable to snippet and notice the sync there happens when all variables read out the leader's stored value.  There needs to be a sync point here before the read.

Reply:
if that is the case shall we move to something like this?
`                        int active_radius;`
`                        //broadcast warp_radius`
`                        if (active_lane == laneId)`
`                               active_radius = __float_as_int(radius);`
`                        radius = __int_as_float(__shfl_sync(0xFFFFFFFF, active_radius, active_lane));
`

Reply:
Bikeshedding: @haritha-j spend some time reading how to use markdown properly on GitHub.
https://guides.github.com/features/mastering-markdown/
 

1. What's the point of reinterpreting back and forth between `int` and `float` if you're not storing the value in the original shared memory block?
2. For now I would keep writing things in `storage.per_warp_buffer[warpId]` as well, like the original implementation was doing.

Reply:
> Bikeshedding: @haritha-j spend some time reading how to use markdown properly on GitHub.

will do :)

> What's the point of reinterpreting back and forth between int and float if you're not storing the value in the original shared memory block?

Sorry, this was because I was under the (incorrect) impression that the variable for `__shffl_sync` had to be an `unisgned_int`. I guess my previous suggestion boils down to 
`radius = __shfl_sync(0xFFFFFFFF, radius, active_lane);`

Wouldn't this be a better approach than previous method of using the ``per_warp_buffer` and not being assured of synchronization?

On the original issue #3583 , @Cazadorro raises some concerns regarding this approach as well

> as a way to check on the state of all warps within the current block, but with out block synchronization anywhere, I don't see how this is happening reliably. If that isn't the goal, I don't see why there needs to be a value in share memory taken put by warps, and is probably a performance concern in the first place.

Reply:
> Wouldn't this be a better approach than previous method of using the ``per_warp_buffer` and not being assured of synchronization?

Yes and your and @Cazadorro concerns are totally valid. This section looks bug filled.

Reply:
So it needed a dive into Nvidia's memory architecture: we're using deprecated functionality of Nvidia where volatile memory was treated as a replacement for atomic, since they lacked atomic instrinsics. Since then, they've updated their drivers and language model to add `std::atomic` explicitly, and don't recommend depending on `volatile` doing the trick.

Original impl makes sense after this fact, because I've used such tricks on low-level code before. But this creates bi-directional memory barriers on every operation and is a performance bottleneck. The proposed `__shfl_sync` is a **much** better method since it only inserts one-directional barrier.

Reply:
So should I proceed with that method for just `radius` or should we also use the same method to get `query`? Or would it be better to leave it as is?

Reply:
> So it needed a dive into Nvidia's memory architecture: we're using deprecated functionality of Nvidia where volatile memory was treated as a replacement for atomic, since they lacked atomic instrinsics.

Can you provide a source for this? The search results I'm getting even within the context of CUDA simply discuss the traditional `volatile` definition.

Reply:
Found [this presentation](https://on-demand.gputechconf.com/supercomputing/2019/pdf/sc1942-the-cuda-c++-standard-library.pdf) as a digestible source.